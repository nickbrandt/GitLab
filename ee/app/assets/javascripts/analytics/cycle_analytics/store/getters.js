import dateFormat from 'dateformat';
import { isNumber } from 'lodash';
import { dateFormats } from '~/analytics/shared/constants';
import { OVERVIEW_STAGE_ID } from '~/cycle_analytics/constants';
import { pathNavigationData as basePathNavigationData } from '~/cycle_analytics/store/getters';
import { filterStagesByHiddenStatus } from '~/cycle_analytics/utils';
import { getIdFromGraphQLId } from '~/graphql_shared/utils';
import httpStatus from '~/lib/utils/http_status';
import { filterToQueryObject } from '~/vue_shared/components/filtered_search_bar/filtered_search_utils';
import { DEFAULT_VALUE_STREAM_ID, OVERVIEW_STAGE_CONFIG, PAGINATION_TYPE } from '../constants';

export const hasNoAccessError = (state) => state.errorCode === httpStatus.FORBIDDEN;

export const currentValueStreamId = ({ selectedValueStream }) =>
  selectedValueStream?.id || DEFAULT_VALUE_STREAM_ID;

export const currentGroupPath = ({ currentGroup }) => currentGroup?.fullPath || null;

export const selectedProjectIds = ({ selectedProjects }) =>
  selectedProjects?.map(({ id }) => getIdFromGraphQLId(id)) || [];

export const cycleAnalyticsRequestParams = (state, getters) => {
  const {
    createdAfter = null,
    createdBefore = null,
    filters: {
      authors: { selected: selectedAuthor },
      milestones: { selected: selectedMilestone },
      assignees: { selectedList: selectedAssigneeList },
      labels: { selectedList: selectedLabelList },
    },
  } = state;

  const filterBarQuery = filterToQueryObject({
    milestone_title: selectedMilestone,
    author_username: selectedAuthor,
    label_name: selectedLabelList,
    assignee_username: selectedAssigneeList,
  });

  return {
    project_ids: getters.selectedProjectIds,
    created_after: createdAfter ? dateFormat(createdAfter, dateFormats.isoDate) : null,
    created_before: createdBefore ? dateFormat(createdBefore, dateFormats.isoDate) : null,
    ...filterBarQuery,
  };
};

export const paginationParams = ({ pagination: { page, sort, direction } }) => ({
  pagination: PAGINATION_TYPE,
  sort,
  direction,
  page,
});

export const hiddenStages = ({ stages }) => filterStagesByHiddenStatus(stages);
export const activeStages = ({ stages }) => filterStagesByHiddenStatus(stages, false);

export const enableCustomOrdering = ({ stages, errorSavingStageOrder }) =>
  stages.some((stage) => isNumber(stage.id)) && !errorSavingStageOrder;

export const customStageFormActive = ({ isCreatingCustomStage, isEditingCustomStage }) =>
  Boolean(isCreatingCustomStage || isEditingCustomStage);

export const isOverviewStageSelected = ({ selectedStage }) =>
  selectedStage?.id === OVERVIEW_STAGE_ID;

/**
 * Until there are controls in place to edit stages outside of the stage table,
 * the path navigation component will only display active stages.
 *
 * https://gitlab.com/gitlab-org/gitlab/-/issues/216227
 */
export const pathNavigationData = ({ stages, medians, stageCounts, selectedStage }) =>
  basePathNavigationData({
    stages: [OVERVIEW_STAGE_CONFIG, ...stages],
    medians,
    stageCounts,
    selectedStage,
  });

export const selectedStageCount = ({ selectedStage, stageCounts }) =>
  stageCounts[selectedStage.id] || null;
