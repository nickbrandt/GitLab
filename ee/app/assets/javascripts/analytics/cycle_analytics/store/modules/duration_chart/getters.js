import { getDurationChartData } from '../../../utils';

export const durationChartPlottableData = (state, _, rootState) => {
  const { createdAfter, createdBefore } = rootState;
  const { durationData } = state;
  const selectedStagesDurationData = durationData.filter((stage) => stage.selected);
  const plottableData = getDurationChartData(
    selectedStagesDurationData,
    createdAfter,
    createdBefore,
  );

  return plottableData.length ? plottableData : [];
};
