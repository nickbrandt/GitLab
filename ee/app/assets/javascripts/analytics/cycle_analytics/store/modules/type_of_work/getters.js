import { getTasksByTypeData } from '../../../utils';

export const selectedTasksByTypeFilters = (state = {}, _, rootState = {}) => {
  const { selectedLabelIds = [], subject } = state;
  const {
    currentGroup,
    selectedProjectIds = [],
    createdAfter = null,
    createdBefore = null,
  } = rootState;
  return {
    currentGroup,
    selectedProjectIds,
    createdAfter,
    createdBefore,
    selectedLabelIds,
    subject,
  };
};

export const tasksByTypeChartData = ({ data = [] } = {}, _, rootState = {}) => {
  const { createdAfter = null, createdBefore = null } = rootState;
  return data.length
    ? getTasksByTypeData({
        data,
        startDate: createdAfter,
        endDate: createdBefore,
      })
    : { groupBy: [], data: [] };
};
