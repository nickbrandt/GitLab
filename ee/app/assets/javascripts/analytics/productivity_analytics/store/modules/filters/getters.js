import dateFormat from 'dateformat';
import { dateFormats } from '~/analytics/shared/constants';
import { getDateInPast, beginOfDayTime, endOfDayTime } from '~/lib/utils/datetime_utility';
import { chartKeys, scatterPlotAddonQueryDays } from '../../../constants';

/**
 * Returns an object of common filter parameters based on the filter's state
 * which will be used for querying the API to retrieve chart and MR table data.
 * The returned object hast the following form:
 *
 * {
 *   group_id: 'gitlab-org',
 *   project_id: 'gitlab-org/gitlab-test',
 *   author_username: 'author',
 *   milestone_title: 'my milestone',
 *   label_name: ['my label', 'yet another label'],
 *   merged_after: '2019-06-11T00:00:00Z'
 *   merged_before: '2019-09-09T23:59:59Z'
 * }
 *
 */
export const getCommonFilterParams = (state, getters) => (chartKey) => {
  const {
    groupNamespace,
    projectPath,
    startDate,
    endDate,
    authorUsername,
    labelName,
    milestoneTitle,
  } = state;

  // for the scatterplot we need to query the API with a date prior to the selected start date
  const mergedAfterDate =
    chartKey && chartKey === chartKeys.scatterplot
      ? dateFormat(getters.scatterplotStartDate, dateFormats.isoDate)
      : dateFormat(startDate, dateFormats.isoDate);

  return {
    group_id: groupNamespace,
    project_id: projectPath,
    author_username: authorUsername,
    milestone_title: milestoneTitle,
    label_name: labelName,
    merged_after: `${mergedAfterDate}${beginOfDayTime}`,
    merged_before: `${dateFormat(endDate, dateFormats.isoDate)}${endOfDayTime}`,
  };
};

/**
 * Returns the start date for the scatterplot.
 * It computes a dateInPast based on the selected startDate
 * and a default number of offset days (offsetDays)
 *
 * When a minDate exists and the minDate is after the computed dateInPast,
 * the minDate is returned. Otherwise the computed dateInPast is returned.
 */
export const scatterplotStartDate = (state) => {
  const { startDate, minDate } = state;
  const dateInPast = getDateInPast(startDate, scatterPlotAddonQueryDays);

  return minDate && minDate > dateInPast ? minDate : dateInPast;
};
