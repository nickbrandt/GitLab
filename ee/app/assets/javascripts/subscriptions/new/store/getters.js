import { s__ } from '~/locale';
import { NEW_GROUP } from '../constants';

export const selectedPlanText = (state, getters) => getters.selectedPlanDetails.text;

export const selectedPlanPrice = (state, getters) =>
  getters.selectedPlanDetails.pricePerUserPerYear;

export const selectedPlanDetails = (state) =>
  state.availablePlans.find((plan) => plan.value === state.selectedPlan);

export const confirmOrderParams = (state, getters) => ({
  setup_for_company: state.isSetupForCompany,
  selected_group: getters.selectedGroupId,
  new_user: state.isNewUser,
  customer: {
    country: state.country,
    address_1: state.streetAddressLine1,
    address_2: state.streetAddressLine2,
    city: state.city,
    state: state.countryState,
    zip_code: state.zipCode,
    company: state.organizationName,
  },
  subscription: {
    plan_id: state.selectedPlan,
    payment_method_id: state.paymentMethodId,
    quantity: state.numberOfUsers,
    source: state.source,
  },
});

export const endDate = (state) =>
  new Date(state.startDate).setFullYear(state.startDate.getFullYear() + 1);

export const totalExVat = (state, getters) => state.numberOfUsers * getters.selectedPlanPrice;

export const vat = (state, getters) => state.taxRate * getters.totalExVat;

export const totalAmount = (_, getters) => getters.totalExVat + getters.vat;

export const name = (state, getters) => {
  if (state.isSetupForCompany && state.organizationName) {
    return state.organizationName;
  } else if (getters.isGroupSelected && getters.isSelectedGroupPresent) {
    return getters.selectedGroupName;
  } else if (state.isSetupForCompany) {
    return s__('Checkout|Your organization');
  }

  return state.fullName;
};

export const usersPresent = (state) => state.numberOfUsers > 0;

export const isGroupSelected = (state) =>
  state.selectedGroup !== null && state.selectedGroup !== NEW_GROUP;

export const isSelectedGroupPresent = (state, getters) => {
  return (
    getters.isGroupSelected && state.groupData.some((group) => group.value === state.selectedGroup)
  );
};

export const selectedGroupUsers = (state, getters) => {
  if (!getters.isGroupSelected) {
    return 1;
  } else if (getters.isSelectedGroupPresent) {
    return state.groupData.find((group) => group.value === state.selectedGroup).numberOfUsers;
  }

  return null;
};

export const selectedGroupName = (state, getters) => {
  if (!getters.isGroupSelected) return null;
  return state.groupData.find((group) => group.value === state.selectedGroup).text;
};

export const selectedGroupId = (state, getters) =>
  getters.isGroupSelected ? state.selectedGroup : null;
