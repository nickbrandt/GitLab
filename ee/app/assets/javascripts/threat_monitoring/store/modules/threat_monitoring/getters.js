import { INVALID_CURRENT_ENVIRONMENT_NAME } from '../../../constants';

export const currentEnvironment = ({ currentEnvironmentId, environments }) =>
  environments.find((environment) => environment.id === currentEnvironmentId);

export const currentEnvironmentName = (state, getters) =>
  getters.currentEnvironment?.name ?? INVALID_CURRENT_ENVIRONMENT_NAME;

export const currentEnvironmentGid = (state, getters) => getters.currentEnvironment?.global_id;

export const canChangeEnvironment = ({
  isLoadingEnvironments,
  isLoadingNetworkPolicyStatistics,
  environments,
}) => !isLoadingEnvironments && !isLoadingNetworkPolicyStatistics && environments.length > 0;
