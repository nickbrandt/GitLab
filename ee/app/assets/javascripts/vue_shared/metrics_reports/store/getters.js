import { LOADING, ERROR, SUCCESS } from '../constants';

export const summaryStatus = (state) => {
  if (state.isLoading) {
    return LOADING;
  }

  if (state.hasError || state.numberOfChanges > 0) {
    return ERROR;
  }

  return SUCCESS;
};

export const metrics = (state) => [
  ...state.changedMetrics,
  ...state.newMetrics.map((metric) => ({ ...metric, isNew: true })),
  ...state.removedMetrics.map((metric) => ({ ...metric, wasRemoved: true })),
  ...state.unchangedMetrics,
];
