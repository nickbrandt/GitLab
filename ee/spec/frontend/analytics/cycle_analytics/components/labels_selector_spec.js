import { GlDropdownSectionHeader } from '@gitlab/ui';
import { mount, shallowMount } from '@vue/test-utils';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import Vue from 'vue';
import Vuex from 'vuex';
import LabelsSelector from 'ee/analytics/cycle_analytics/components/labels_selector.vue';
import createStore from 'ee/analytics/cycle_analytics/store';
import * as getters from 'ee/analytics/cycle_analytics/store/getters';
import waitForPromises from 'helpers/wait_for_promises';
import createFlash from '~/flash';
import { groupLabels } from '../mock_data';

jest.mock('~/flash');
Vue.use(Vuex);

const selectedLabel = groupLabels[groupLabels.length - 1];
const findActiveItem = (wrapper) =>
  wrapper
    .findAll('gl-dropdown-item-stub')
    .filter((d) => d.attributes('active'))
    .at(0);

const mockGroupLabelsRequest = (status = 200) =>
  new MockAdapter(axios).onGet().reply(status, groupLabels);

describe('Value Stream Analytics LabelsSelector', () => {
  let store = null;

  function createComponent({ props = { selectedLabelIds: [] }, shallow = true } = {}) {
    store = createStore();
    const func = shallow ? shallowMount : mount;
    return func(LabelsSelector, {
      store: {
        ...store,
        getters: {
          ...getters,
          currentGroupPath: 'fake',
        },
      },
      propsData: {
        ...props,
      },
    });
  }

  let wrapper = null;
  let mock = null;
  const labelNames = groupLabels.map(({ name }) => name);

  describe('with no item selected', () => {
    beforeEach(() => {
      mock = mockGroupLabelsRequest();
      wrapper = createComponent({});

      return waitForPromises();
    });

    afterEach(() => {
      mock.restore();
      wrapper.destroy();
      wrapper = null;
    });

    it('will render the label selector', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it.each(labelNames)('generate a label item for the label %s', (name) => {
      expect(wrapper.text()).toContain(name);
    });

    it('will fetch the labels', () => {
      expect(mock.history.get.length).toBe(1);
    });

    it('will render with the default option selected', () => {
      const sectionHeader = wrapper.findComponent(GlDropdownSectionHeader);

      expect(sectionHeader.exists()).toBe(true);
      expect(sectionHeader.text()).toEqual('Select a label');
    });

    describe('with a failed request', () => {
      beforeEach(() => {
        mock = mockGroupLabelsRequest(404);
        wrapper = createComponent({});

        return waitForPromises();
      });

      it('should flash an error message', () => {
        expect(createFlash).toHaveBeenCalledWith({
          message: 'There was an error fetching label data for the selected group',
        });
      });
    });

    describe('when a dropdown item is clicked', () => {
      beforeEach(() => {
        mock = mockGroupLabelsRequest();
        wrapper = createComponent({ shallow: false });
        return waitForPromises();
      });

      it('will emit the "select-label" event', () => {
        expect(wrapper.emitted('select-label')).toBeUndefined();

        const elem = wrapper.findAll('.dropdown-item').at(1);
        elem.trigger('click');

        return wrapper.vm.$nextTick().then(() => {
          expect(wrapper.emitted('select-label').length > 0).toBe(true);
          expect(wrapper.emitted('select-label')[0]).toContain(groupLabels[1].id);
        });
      });
    });
  });

  describe('with selectedLabelIds set', () => {
    beforeEach(() => {
      mock = mockGroupLabelsRequest();
      wrapper = createComponent({ props: { selectedLabelIds: [selectedLabel.id] } });
      return waitForPromises();
    });

    afterEach(() => {
      wrapper.destroy();
    });

    it('will render the label selector', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('will set the active label', () => {
      const activeItem = findActiveItem(wrapper);

      expect(activeItem.exists()).toBe(true);
      expect(activeItem.text()).toEqual(selectedLabel.name);
    });
  });

  describe('with labels provided', () => {
    beforeEach(() => {
      mock = mockGroupLabelsRequest();
      wrapper = createComponent({ props: { initialData: groupLabels } });
    });

    afterEach(() => {
      wrapper.destroy();
    });

    it('will not fetch the labels', () => {
      expect(mock.history.get.length).toBe(0);
    });
  });
});
