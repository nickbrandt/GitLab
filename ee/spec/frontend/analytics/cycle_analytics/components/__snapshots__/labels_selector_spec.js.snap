// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Value Stream Analytics LabelsSelector with no item selected will render the label selector 1`] = `
"<gl-dropdown-stub headertext=\\"\\" hideheaderborder=\\"true\\" text=\\"\\" category=\\"primary\\" variant=\\"default\\" size=\\"medium\\" toggleclass=\\"gl-overflow-hidden\\" class=\\"gl-w-full\\">
  <gl-dropdown-section-header-stub>Select a label </gl-dropdown-section-header-stub>
  <div class=\\"mb-3 px-3\\">
    <gl-search-box-by-type-stub value=\\"\\" clearbuttontitle=\\"Clear\\"></gl-search-box-by-type-stub>
  </div>
  <div class=\\"mb-3 px-3\\">
    <gl-dropdown-item-stub avatarurl=\\"\\" iconcolor=\\"\\" iconname=\\"\\" iconrightarialabel=\\"\\" iconrightname=\\"\\" secondarytext=\\"\\" class=\\"\\">
      <!----> <span class=\\"d-inline-block dropdown-label-box\\" style=\\"background-color: rgb(255, 0, 0);\\"></span>
      roses
    </gl-dropdown-item-stub>
    <gl-dropdown-item-stub avatarurl=\\"\\" iconcolor=\\"\\" iconname=\\"\\" iconrightarialabel=\\"\\" iconrightname=\\"\\" secondarytext=\\"\\" class=\\"\\">
      <!----> <span class=\\"d-inline-block dropdown-label-box\\" style=\\"background-color: rgb(255, 255, 255);\\"></span>
      some space
    </gl-dropdown-item-stub>
    <gl-dropdown-item-stub avatarurl=\\"\\" iconcolor=\\"\\" iconname=\\"\\" iconrightarialabel=\\"\\" iconrightname=\\"\\" secondarytext=\\"\\" class=\\"\\">
      <!----> <span class=\\"d-inline-block dropdown-label-box\\" style=\\"background-color: rgb(0, 0, 255);\\"></span>
      violets
    </gl-dropdown-item-stub>
    <div class=\\"text-center\\" style=\\"display: none;\\">
      <gl-loading-icon-stub label=\\"Loading\\" size=\\"md\\" color=\\"dark\\" inline=\\"true\\"></gl-loading-icon-stub>
    </div>
    <div class=\\"text-secondary\\" style=\\"display: none;\\">
      No matching labels
    </div>
  </div>
</gl-dropdown-stub>"
`;

exports[`Value Stream Analytics LabelsSelector with selectedLabelIds set will render the label selector 1`] = `
"<gl-dropdown-stub headertext=\\"\\" hideheaderborder=\\"true\\" text=\\"\\" category=\\"primary\\" variant=\\"default\\" size=\\"medium\\" toggleclass=\\"gl-overflow-hidden\\" class=\\"gl-w-full\\">
  <gl-dropdown-section-header-stub>Select a label </gl-dropdown-section-header-stub>
  <div class=\\"mb-3 px-3\\">
    <gl-search-box-by-type-stub value=\\"\\" clearbuttontitle=\\"Clear\\"></gl-search-box-by-type-stub>
  </div>
  <div class=\\"mb-3 px-3\\">
    <gl-dropdown-item-stub avatarurl=\\"\\" iconcolor=\\"\\" iconname=\\"\\" iconrightarialabel=\\"\\" iconrightname=\\"\\" secondarytext=\\"\\" class=\\"\\">
      <!----> <span class=\\"d-inline-block dropdown-label-box\\" style=\\"background-color: rgb(255, 0, 0);\\"></span>
      roses
    </gl-dropdown-item-stub>
    <gl-dropdown-item-stub avatarurl=\\"\\" iconcolor=\\"\\" iconname=\\"\\" iconrightarialabel=\\"\\" iconrightname=\\"\\" secondarytext=\\"\\" class=\\"\\">
      <!----> <span class=\\"d-inline-block dropdown-label-box\\" style=\\"background-color: rgb(255, 255, 255);\\"></span>
      some space
    </gl-dropdown-item-stub>
    <gl-dropdown-item-stub avatarurl=\\"\\" iconcolor=\\"\\" iconname=\\"\\" iconrightarialabel=\\"\\" iconrightname=\\"\\" secondarytext=\\"\\" active=\\"true\\" class=\\"\\">
      <!----> <span class=\\"d-inline-block dropdown-label-box\\" style=\\"background-color: rgb(0, 0, 255);\\"></span>
      violets
    </gl-dropdown-item-stub>
    <div class=\\"text-center\\" style=\\"display: none;\\">
      <gl-loading-icon-stub label=\\"Loading\\" size=\\"md\\" color=\\"dark\\" inline=\\"true\\"></gl-loading-icon-stub>
    </div>
    <div class=\\"text-secondary\\" style=\\"display: none;\\">
      No matching labels
    </div>
  </div>
</gl-dropdown-stub>"
`;
